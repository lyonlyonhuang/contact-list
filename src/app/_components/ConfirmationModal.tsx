import React ,{ useEffect } from 'react';
import { createPortal } from 'react-dom';
import './layout.css'

interface PropsType {
  uiqID: string
  text:string, 
  open:boolean,
  buttonColor: {
    cancelButton: string,
    confirmButton: string
  },
  handleConfirm:(arg:boolean, event?:any) => void,
  setConfirmModalOpen: (arg:boolean) => void
}

const ConfirmationModal = ( {uiqID, text, open, buttonColor, handleConfirm, setConfirmModalOpen}: PropsType ) => {
  
  useEffect(() => {
    const timerId = window.setTimeout(() => {
      setConfirmModalOpen(false);
    }, 55000);
    if (open) {
        timerId
    }
    return () => {
      window.clearTimeout(timerId);
    }
  }, [open]);
  
  return (
    open ?
    createPortal(
      <div data-testid='confirmation-modal'>
        <div className='modal '>
          <button onClick={() => handleConfirm(false)} className='modal__close'>&times;</button>
          <div className="confirm-content">
              <h4>CONFIRMATION!</h4>
              <div>
                <h2>{text}</h2>
                <p>Please click confirm or cancel button to continue</p>
              </div>
            </div>
            <div className="">
              <button data-testid='confirmation-cancel-button' className={`w-30 p-2 m-1 border-2 rounded-md md:w-36 ${buttonColor.cancelButton}`} onClick={() => handleConfirm(false)}>Cancel</button>
              <button data-testid='confirmation-confirm-button' className={`w-30 p-2 m-1 border-2 rounded-md md:w-36 text-white  ${buttonColor.confirmButton}`} onClick={() => handleConfirm(true)}>Confirm</button>
            </div>
        </div>
        <div 
          className="overlay" 
          onClick={() => handleConfirm(false)} 
        />
      </div>
      ,
    document.body, `confirmation-modal-${uiqID}`)
    : null
  )
}

  export default ConfirmationModal