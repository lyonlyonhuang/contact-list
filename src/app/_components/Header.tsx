import React from 'react';
import Link from 'next/link'

export default function Header () {
  return (
    <div className='text-center py-4 border-gray-400 border-t-2 border-b-2 md:text-start md:pl-5'>
      <Link href={`/contact`} >
            Contact List
      </Link>
    </div>
  );
}
