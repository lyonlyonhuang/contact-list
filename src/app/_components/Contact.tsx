import React, { useState, useEffect } from 'react'
import { ContactModel } from '../_types_interface'
import Link from 'next/link'
import Image from "next/image";
import DeleteImg from '../../../public/delete.svg'
import EditImg from '../../../public/edit.svg'
import ConfirmationModal from './ConfirmationModal'

const DeleteColor = {
  cancelButton: 'border-2 border-red-500 text-red-500 hover:bg-red-100',
  confirmButton: 'border-2 border-red-500 bg-red-500 hover:bg-red-400'
}

export default function Contact(params: ContactModel) {
  const [confirmModalOpen, setConfirmModalOpen] = useState<boolean>(false)
 
  const popupConfirmModal = () => {
    setConfirmModalOpen(true);
  };

  const handleConfirm = (confirm:boolean) => {
      if(confirm) {
          params.deleteContact(params.id)
      }
      setConfirmModalOpen(false)
  }

  return (
    <div className='p-1 mx-2 my-1 border-gray-400 border-2 rounded flex bg-slate-50 hover:bg-slate-100'>
	      <div className='flex shrink-0 items-center m-1 lg:m-4'>
            <img className='rounded-full w-16 h-16 lg:w-32 lg:h-32' src="https://i.pravatar.cc/300" alt="mock avatar" />
        </div>
        <div className='flex flex-col grow m-1 pr-1 lg:m-4'>
            <div className='flex items-center'>
              <div className='text-sm font-bold mr-2 lg:text-xl'>{params.first_name} {params.last_name} </div> 
              <>|</>
              <div className='text-sm ml-2 lg:text-xl'>{params.job}</div>
            </div>
            <div className='text-xs lg:text-sm'>{params.description}</div>
        </div>
        <div className='flex shrink-0 items-center'>
            <Link data-testid='contact-edit-link' href={`/contact/edit/${params.id}`} className='right-3'>
                <Image data-testid='contact-edit-button' className='w-6 h-6 m-2 lg:w-9 lg:h-9 lg:m-4' src={EditImg} alt='Edit' />
            </Link>
            <div onClick={() => popupConfirmModal()}>
                <Image data-testid='contact-delete-button' className='w-6 h-6 m-2 lg:w-9 lg:h-9 lg:m-4 cursor-pointer' src={DeleteImg} alt='Delete' />
            </div>
            <ConfirmationModal 
              text={'Delete contact'} 
              open={confirmModalOpen} 
              buttonColor={DeleteColor} 
              uiqID={`${Math.random()}`}
              handleConfirm={handleConfirm}
              setConfirmModalOpen={setConfirmModalOpen}
              />
        </div>
    </div>
  )
}