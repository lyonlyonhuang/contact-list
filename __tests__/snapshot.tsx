import { render, screen } from '@testing-library/react'
import Home from '@/app/page'
import Contacts from '@/app/contact/page'
import Contact from '@/app/_components/Contact'
import Header from '@/app/_components/Header'
import ConfirmationModal from '@/app/_components/ConfirmationModal'

describe('Home', () => {
    it('renders homepage unchanged', () => {
        const { container } = render(<Home />)
        expect(container).toMatchSnapshot()
    })
})

describe('Contacts', () => {
    it('renders Contacts unchanged', () => {
        const { container } = render(<Contacts />)
        expect(container).toMatchSnapshot()
    })
})

describe('Contact', () => {
    const mockContact = {
        id: 1,
        first_name: 'Alex',
        last_name: 'Wang',
        job: 'PM',
        description: 'this is description',
        deleteContact:() => jest.fn() 
    }
    it('renders Contact unchanged', () => {
        const { container } = render(<Contact {...mockContact}/>)
        expect(container).toMatchSnapshot()
    })
})

describe('Header', () => {
    it('renders Header unchanged', () => {
        const { container } = render(<Header />)
        expect(container).toMatchSnapshot()
    })
})

describe('ConfirmationModal', () => {
    const mockModal = {
        uiqID: '1',
        text: 'Delete', 
        open:true,
        buttonColor: {
          cancelButton: 'border-2 border-red-500 text-red-500 hover:bg-red-100',
          confirmButton: 'border-2 border-red-500 bg-red-500 hover:bg-red-400'
        },
        handleConfirm:() => jest.fn() ,
        setConfirmModalOpen: () => jest.fn() 
    }
    it('renders ConfirmationModal unchanged', () => {
        const { container } = render(<ConfirmationModal 
            text={mockModal.text} 
            open={mockModal.open} 
            buttonColor={mockModal.buttonColor} 
            uiqID={mockModal.uiqID}
            handleConfirm={mockModal.handleConfirm}
            setConfirmModalOpen={mockModal.setConfirmModalOpen}
        />)
        expect(container).toMatchSnapshot()
    })
})