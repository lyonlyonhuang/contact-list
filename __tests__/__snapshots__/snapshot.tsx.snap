// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmationModal renders ConfirmationModal unchanged 1`] = `<div />`;

exports[`Contact renders Contact unchanged 1`] = `
<div>
  <div
    class="p-1 mx-2 my-1 border-gray-400 border-2 rounded flex bg-slate-50 hover:bg-slate-100"
  >
    <div
      class="flex shrink-0 items-center m-1 lg:m-4"
    >
      <img
        alt="mock avatar"
        class="rounded-full w-16 h-16 lg:w-32 lg:h-32"
        src="https://i.pravatar.cc/300"
      />
    </div>
    <div
      class="flex flex-col grow m-1 pr-1 lg:m-4"
    >
      <div
        class="flex items-center"
      >
        <div
          class="text-sm font-bold mr-2 lg:text-xl"
        >
          Alex
           
          Wang
           
        </div>
        |
        <div
          class="text-sm ml-2 lg:text-xl"
        >
          PM
        </div>
      </div>
      <div
        class="text-xs lg:text-sm"
      >
        this is description
      </div>
    </div>
    <div
      class="flex shrink-0 items-center"
    >
      <a
        class="right-3"
        data-testid="contact-edit-link"
        href="/contact/edit/1"
      >
        <img
          alt="Edit"
          class="w-6 h-6 m-2 lg:w-9 lg:h-9 lg:m-4"
          data-nimg="1"
          data-testid="contact-edit-button"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
      </a>
      <div>
        <img
          alt="Delete"
          class="w-6 h-6 m-2 lg:w-9 lg:h-9 lg:m-4 cursor-pointer"
          data-nimg="1"
          data-testid="contact-delete-button"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Contacts renders Contacts unchanged 1`] = `
<div>
  <div>
    Failed to load
  </div>
</div>
`;

exports[`Header renders Header unchanged 1`] = `
<div>
  <div
    class="text-center py-4 bg-yellow-400  md:text-start md:pl-5"
  >
    <a
      data-testid="header-link"
      href="/contact"
    >
      <div
        class="text-slate-50 text-xl font-black w-fit inline-block md:block"
        id="header"
      >
        Contact List
      </div>
    </a>
  </div>
</div>
`;

exports[`Home renders homepage unchanged 1`] = `
<div>
  <main>
    <div>
      <div>
        Failed to load
      </div>
    </div>
  </main>
</div>
`;
